File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = INT_MIN;

    int maxPathSum(TreeNode* root) {
        dfs(root);
        
        return ans;
    }
    
    int dfs(TreeNode* root) {
        if(root == nullptr) {
            return -10000;
        }

        int l = dfs(root->left);
        int r = dfs(root->right);

        ans = max(ans, root->val);
        ans = max(ans, root->val+l);
        ans = max(ans, root->val+r);
        ans = max(ans, root->val+l+r);

        int curr = root->val;
        curr = max(curr, root->val+l);
        curr = max(curr, root->val+r);

        return curr;
    }
};

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        dfs(root);
        return ans;
    }

    private int dfs(TreeNode root) {
        if(root == null) {
            return -10000;
        }

        int l = dfs(root.left);
        int r = dfs(root.right);

        ans = Math.max(ans, root.val);
        ans = Math.max(ans, root.val+l);
        ans = Math.max(ans, root.val+r);
        ans = Math.max(ans, root.val+r+l);

        int curr = root.val;
        curr = Math.max(curr, root.val+l);
        curr = Math.max(curr, root.val+r);

        return curr;
    }
}